napari/_qt/_tests/test_plugin_widgets.py

```diff
@@ -195,3 +133,67 @@ def test_inject_viewer_proxy(make_napari_viewer):
     with patch('napari.utils.misc.ROOT_DIR', new='/some/other/package'):
         with pytest.warns(FutureWarning):
             wdg.fail()
+
+
+@pytest.mark.parametrize(
+    "widget_callable, param",
+    [
+        (QWidget_example, 'napari_viewer'),
+        (QWidget_string_annnot, 'test'),
+        (Container_example, 'test'),
+    ],
+)
+def test_get_widget_viewer_param(widget_callable, param):
+    """Test `_get_widget_viewer_param` returns correct parameter name."""
+    out = _get_widget_viewer_param(widget_callable, 'widget_name')
+    assert out == param
+
+
+def test_get_widget_viewer_param_error():
+    """Test incorrect subclass raises error in `_get_widget_viewer_param`."""
+    with pytest.raises(TypeError) as e:
+        _get_widget_viewer_param(ErrorWidget, 'widget_name')
+    assert "'widget_name' must be `QtWidgets.QWidget`" in str(e)
+
+
+def test_widget_hide_destroy(make_napari_viewer):
+    """Test that widget hide and destroy works."""
+    viewer = make_napari_viewer()
+    viewer.window.add_dock_widget(QWidget_example(viewer), name='test')
+    widget = viewer.window._dock_widgets['test']
+
+    # Check widget persists after hide
+    widget.title.hide_button.click()
+    assert widget
+    # Check that widget removed from `_dock_widgets` dict when closed
+    widget.destroyOnClose()
+    assert 'test' not in viewer.window._dock_widgets
+
+
+@pytest.mark.parametrize(
+    "Widget",
+    [
+        QWidget_example,
+        Container_example,
+        magic_widget_example,
+        callable_example,
+    ],
+)
+def test_widget_types_supported(
+    make_napari_viewer,
+    tmp_plugin: DynamicPlugin,
+    Widget,
+):
+    """Test all supported widget types correctly instantiated and call processor."""
+    # The 4 `Widget`s should represent varing widget constructors and signatures that
+    # we want to support.
+    tmp_plugin.contribute.widget(display_name='Widget')(Widget)
```

Comment:
@lucyleeow sorry, I'm being a bit slow but can you please explain this syntax? I didn't get it originally when Ashley proposed it and I still don't hahaha.

What's the return value of `tmp_plugin.contribute.widget(display_name='Widget')` and how does calling it with `(Widget)` as a parameter achieve what we want (which, if I got it right, is assigning `Widget` as a callable for the widget contribution we've just created).

Reply:
I can try to explain my thinking here!

A decorator is a callable that takes a callable and returns a callable. If you use the `@`-syntax it is applied to whatever is defined just under it. In other words, this:
```
@gday
def hello():
    print("Hello, world!")
```
is equivalent to:
```
def hello():
    print("Hello, world!")

hello = gday(hello)
```

Usually users making a contribution will use it where they define the fn (or class) that generates their widget, so they can use the `@`-syntax right above.

In the test, the callables (some classes, some functions) that construct the widgets are defined elsewhere, so I proposed calling the decorator on the parametrized widget-returning-callable (`Widget` in the test) using normal function syntax. The wrapper function with decorator in the test before was an extra layer interfering with the process.

Part of the reason this ends up pretty weird looking is that this is a decorator with arguments, so you end up with that double-call `foo(bar)(baz)` syntax.

Reply:
Yep yep, got it. Thanks for the detailed explanation! Might be worth a comment, but I'm not too concerned

Reply:
I will add a comment, and I'll mention why it is there (to prevent widget leaks)